2 new events (AllowPlayerClaimPermissionEvent, PlayerAbandonClaimEvent) - bis

Vilevar opened the pull request:
Because the IAddonPlugin interface of EtienneDx isn't on the code and I didn't find how I can get the same skills with what there already is, I created 2 new events :  
- AllowPlayerClaimPermissionEvent : It's fired when a method "allowSomething" in the class Claim is called with the parameter "fireEvent" on true (so we also can have without event to avoid loops) and it allows to modify the denial message. The only "special" thing is the permission type (from the enum AllowedPermissionType) which depends on which method from the class Claim called the event (allowEdit or allowBuild or ...) so there is one event for all permission types (a little bit as PlayerInteractEvent for all actions).
- PlayerAbandonClaimEvent : It's fired when a player try to abadon a claim. I created it because ClaimDeletedEvent is too general and called too late (which implies that it cannot be cancelled).

(I'm not sure if it is 1 or 2 features but because it replaces 1 thing (the IAddonPlugin interface) I send it in a single Pull Request.)

Vilevar committed:
Adding 2 events

Vilevar committed:
Update Claim.java

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -307,51 +311,89 @@ public boolean isNear(Location location, int howNear)
         return claim.contains(location, false, true);
     }
-
+    
```
Shouldn't be trailing whitespace here, only nitpicking it because of other problems.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -307,51 +311,89 @@ public boolean isNear(Location location, int howNear)
-        }
+        String message;
+        // default system before event firing
+        test: {
```
This label usage is pretty bad practice. It makes the code harder to follow and increases indentation for no real benefit - anything done with it could easily be done by adding additional else blocks.

Vilevar replied:
I didn't think about adding "else if" blocks because I wanted to have something as near of the original code as possible but I'm going to fix this (if it's worth it).

Jikoo replied:
I think you're probably just getting too hung up on maintaining a minimal line content diff. When you look at the actual logic, there are few else blocks used, which means that all failed conditions fall through to the next if. That means that if you stack all of the conditions in a single block, you can actually achieve the required behavior with an else.

Examine the following:
```Java
if (condition1) {
  if (condition2) {
    return x;
  }
}
if (condition3) {
  return y;
}
return z;
```
Vs.
```Java
if (condition1 && condition2) {
  return x;
} else if (condition3) {
  return y;
} else {
  return z;
}
```

If it's truly not possible to do in an efficient manner without quick returns, then a helper method should be created instead.

Vilevar replied:
I think I'll use a helper method because there's an `elseif` with the `condition3` so it is not equivalent (maybe in practice well but I don't want to take the risk).

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -307,51 +311,89 @@ public boolean isNear(Location location, int howNear)
+
+    //permissions.  note administrative "public" claims have different rules than other claims
+    //all of these return NULL when a player has permission, or a String error message when the player doesn't have permission
+    public String allowEdit(Player player, boolean fireEvent)
```
Can you explain a usage in which one would not want to fire an event? This does not seem like a good design choice.

This applies to all of the trust/permission checks, not going to bother noting the rest because it's pretty clear where they are.

Vilevar replied:
Yes, when it calls another allowMethod (such as for parent Claim) it should fire twice for the same test which is useless.
I also think that sometimes we might want to have the result of GP only, without the intervention of other plugins.

Jikoo replied:
That's reasonable. It probably should be private then, addons shouldn't be able to bypass checks.

Vilevar replied:
I'm not sure to understand (I'm not English-speaking) : with this system the addons can't bypass the checks, they can "only" modify what's the return of the checks. And with the parameter `fireEvent`, the default checks are also done and I don't think that making this method private is a good idea (e.g. if there is 2 addons, they could create an infinite loop) but if you think it's necessary (to do it private) I'll do.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -367,70 +409,114 @@ private boolean placeableForFarming(Material material)
-        }
+        String message;
+        // default system before event firing
+        test: {
```
Same as above, abusing labels when it's easily avoidable.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -505,125 +591,253 @@ else if (player.getUniqueId().equals(this.ownerID))
+        
+        String message;
+        // default system before event firing
+        test: {
```
More label abuse.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -505,125 +591,253 @@ else if (player.getUniqueId().equals(this.ownerID))
-        }
+        String message;
+        // default system before event firing
+        test: {
```
More label abuse.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -505,125 +591,253 @@ else if (player.getUniqueId().equals(this.ownerID))
-        if (this.allowEdit(player) == null) return null;
+        String message;
+        // default system before event firing
+        test: {
```
More label abuse.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/Claim.java:
```diff
@@ -505,125 +591,253 @@ else if (player.getUniqueId().equals(this.ownerID))
-        return reason;
     }
-
+    
```
Small nitpick again, whitespace.

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/GriefPrevention.java:
```diff
@@ -2923,6 +2924,12 @@ else if (claim.children.size() > 0 && !deleteTopLevelClaim)
+            // fire event
+            PlayerAbandonClaimEvent event = new PlayerAbandonClaimEvent(player, claim);
+            Bukkit.getPluginManager().callEvent(event);
+            if(event.isCancelled())
```
Space after if

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/events/AllowPlayerClaimPermissionEvent.java:
```diff
@@ -0,0 +1,100 @@
+    @Nullable
+    private final Material material;
+
+    public AllowPlayerClaimPermissionEvent(Player player, Claim claim, AllowedPermissionType permissionType, String message) {
```
Braces on new lines for this whole class

Vilevar replied:
I forgot, sorry

Jikoo commented on src/main/java/me/ryanhamshire/GriefPrevention/events/PlayerAbandonClaimEvent.java:
```diff
@@ -0,0 +1,50 @@
+
+import me.ryanhamshire.GriefPrevention.Claim;
+
+public class PlayerAbandonClaimEvent extends Event implements Cancellable {
```
Braces on new lines for this whole class

Vilevar replied:
I didn't know your plans about revisiting claim permission logic and I'm sorry, I think it's a good idea but now it's not like that (I think...) : I only use the duplicate logic which already were there because I've found that the methods `Claim#allowSomething(Player, [Material])` are more "powerful" that the method `Claim#hasExplicitPermission(Player, ClaimPermission)` because these methods use the other one so that means that setting an event at these methods allows us to do more customizations and therefore I did it. I hope you better understand why I worked that way...

About the `PlayerAbandonClaimEvent` (I'll rename it) I didn't think about /abandonall, I didn't look for each ways to unclaim because I was thinking that it was everytime with the same method so I'll fix it.
But I know that ClaimDeletedEvent is better and handles (I think) everything but as I said, it's fired to late, we can't cancel it so I find it not so useful (and personally I need to cancel some abandons)

Jikoo replied:
I assume you meant to respond to the main review, so I'll reply there.

Jikoo reviewed (changes requested):
I want to preface my review by saying that I appreciate how much work goes into changes like this. Thank you for helping out and submitting what you've done.

I would like to see the `PlayerAbandonClaimEvent` ~~(which should actually likely just be an `AbandonClaimEvent` and support any `CommandSender` so it is also fired when console abandons claims)~~ as a separate PR, there's a lot to unpack here. I'm not going to focus on it here due to the separate nature, but I will say that it looks like you're missing a couple cases, at very least `/abandonallclaims`.

Overall I'm not a fan of the route you've taken. There is currently a LOT of duplicate logic across claim checks that could be reduced, and this only adds more. When I added `ClaimPermission#isGrantedBy(ClaimPermission)`, I planned to revisit claim permission logic and refactor at very least all `ClaimPermission` checks into a single method, if not also PermissionTrust.

/e `ClaimDeletedEvent` handles console deleting claims, that's fine. I definitely recall seeing players with a single claim doing `/abandonallclaims` constantly because they'd place a chest temporarily and then move on.

Vilevar committed:
Delete PlayerAbandonClaimEvent + some synthax corrections

Vilevar committed:
allowMethods in handleClaim

The metods Claim#allowEdit, Claim#allowBuild and Claim#allowAccess are not managed by the single method Claim#handleAction (they still exist but are deprecated). For this I added some parameters for the ClaimPermission's objects (I tried to not abuse with that but I also tried to minimize the number of identification of the ClaimPermission), I also added the ClaimPermission "Edit" which is only to fit with the mechanism of Claim#handleAction.
The methods Claim#allowBreak and Claim#allowGrantPermission are not managed in Claim#handleAction because the tests are really different.
The last one also call the event AllowPlayerClaimPermissionEvent so I still use the enum ClaimPermissionType (before it was AllowedPermissionType) because I didn't know if creating an GrantPermission object in the enum ClaimPermission could work.
